scipy/signal/_spectral_py.py

```diff
@@ -134,22 +150,120 @@ def lombscargle(x,
     >>> plt.show()
 
     """
-    x = np.ascontiguousarray(x, dtype=np.float64)
-    y = np.ascontiguousarray(y, dtype=np.float64)
-    freqs = np.ascontiguousarray(freqs, dtype=np.float64)
+
+    # if no weights are provided, assume all data points are equally important
+    if weights is None:
+        weights = np.ones_like(x, dtype=np.float64)
 
     assert x.ndim == 1
     assert y.ndim == 1
     assert freqs.ndim == 1
-
-    if precenter:
-        pgram = _lombscargle(x, y - y.mean(), freqs)
+    assert weights.ndim == 1
+
+    # validate input sizes
+    if x.shape != y.shape or x.shape != weights.shape:
+        raise ValueError("Input arrays do not have the same shape.")
+
+    # check for any freq == 0
+    if np.any(freqs == 0):
+        raise ZeroDivisionError("Freqs array contains a 0.")
+    
+    # check for weights summing to 0
+    if weights.sum() == 0:
+        raise ZeroDivisionError("Weights sum to 0.")
+    
+    # check for any negative weights
+    if np.any(weights < 0):
+        raise ValueError("Each weight must be >= 0.")
+
+    # validate normalize parameter
+    OUTPUT_POWER = "power"
+    OUTPUT_NORMALIZE = "normalize"
+    OUTPUT_AMPLITUDE = "amplitude"
+    if isinstance(normalize, bool):
+        # if bool, convert to str
+        normalize = OUTPUT_NORMALIZE if normalize else OUTPUT_POWER
     else:
-        pgram = _lombscargle(x, y, freqs)
-
-    if normalize:
-        pgram *= 2 / np.dot(y, y)
+        # if neither a bool or str
+        if not isinstance(normalize, str):
+            raise TypeError("Normalize type must a bool or str.")
+
+    if normalize not in [OUTPUT_POWER, OUTPUT_NORMALIZE, OUTPUT_AMPLITUDE]:
+        raise ValueError(
+            "Normalize must be: "
+            f"False (or '{OUTPUT_POWER}'), "
+            f"True (or '{OUTPUT_NORMALIZE}'), "
+            f"or '{OUTPUT_AMPLITUDE}'."
+        )
 
+    # convert inputs to contiguous arrays with high precision
+    x = np.ascontiguousarray(x, dtype=np.float64)
+    y = np.ascontiguousarray(y, dtype=np.float64)
+    freqs = np.ascontiguousarray(freqs, dtype=np.float64)
+    weights = np.ascontiguousarray(weights, dtype=np.float64)
+
+    # weight vector must sum to 1
+    weights = weights / weights.sum()
+
+    # pre-allocate arrays for intermediate and final calculations
+    coswt = np.empty_like(x)
+    sinwt = np.empty_like(x)
+    wcoswt = np.empty_like(x)
+    wsinwt = np.empty_like(x)
```

Comment:
I am not sure, if pre-declaring NumPy arrays bring a measurable speed benefit (though I never did any profiling to verify).

Reply:
I didn't exhaustively test this, but I found a tiny performance impact if those 4 arrays (used only inside the loop) were not preallocated. Given those arrays could be quite large, I think it's safer to error on the side of caution and only allocate them once.